miniapps/parelag/MultilevelHcurlSolver.cpp

```diff
@@ -0,0 +1,739 @@
+
+      // Eliminate the boundary conditions
+      Array<int> marker(spA->Height());
+      marker = 0;
+      sequence[0]->GetDofHandler(jform)->MarkDofsOnSelectedBndr(ess_attr[0],
+                                                                marker);
+
+      for (int i = 0; i < spA->Height(); ++i)
+      {
+         if (marker[i])
+         {
+            spA->EliminateRowCol(i, sol->Elem(i), *rhs);
+         }
+      }
+
+      A = Assemble(hcurl_dofTrueDof, *spA, hcurl_dofTrueDof);
+      hcurl_dofTrueDof.Assemble(*rhs, B);
+   }
+   if (!myid)
+   {
+      cout << "A size: " << A->GetGlobalNumRows() << 'x'
+           << A->GetGlobalNumCols() << '\n' << " A NNZ: " << A->NNZ() << '\n';
+   }
+   MFEM_ASSERT(B.Size() == A->Height(),
+               "Matrix and vector size are incompatible.");
+   assemble_timer.Stop();
+
+   // Perform the solves.
+   Timer solvers_timer = TimeManager::AddTimer("Solvers -- Total");
+   if (!myid)
+   {
+      cout << "\nRunning fine-level solvers...\n\n";
+   }
+
+   // Create the solver library.
+   auto lib = SolverLibrary::CreateLibrary(
+                 master_list->Sublist("Preconditioner Library"));
+
+   // Loop through the solvers.
+   for (const auto& solver_name : list_of_solvers)
+   {
+      Timer solver_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                                 append(solver_name).
+                                                 append("\" -- Total"));
+      // Get the solver factory.
+      auto solver_factory = lib->GetSolverFactory(solver_name);
+      auto solver_state = solver_factory->GetDefaultState();
+      solver_state->SetDeRhamSequence(sequence[0]);
+      solver_state->SetBoundaryLabels(ess_attr);
+      solver_state->SetForms({jform});
+
+      // Build the silver.
+      Timer build_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                                append(solver_name).
+                                                append("\" -- Build"));
+      if (!myid)
+      {
+         cout << "Building solver \"" << solver_name << "\"...\n";
+      }
+      unique_ptr<Solver> solver = solver_factory->BuildSolver(A, *solver_state);
+      build_timer.Stop();
+
+      // Run the solver.
+      Timer pre_timer = TimeManager::AddTimer(std::string("Solver \"").
+                                              append(solver_name).
+                                              append("\" -- Pre-solve"));
+      if (!myid)
+      {
+         cout << "Solving system with \"" << solver_name << "\"...\n";
+      }
+
+      // Note that X is on true dofs owned by the process, while x is on local
+      // dofs that are known to the process.
+      Vector X(A->Width()), x(sequence[0]->GetNumberOfDofs(jform));
+      X=0.0;
+
+      {
+         Vector tmp(A->Height());
+         A->Mult(X, tmp);
+         tmp *= -1.0;
+         tmp += B;
+
+         double local_norm = tmp.Norml2() * tmp.Norml2();
```

Comment:
I'm sure this has no significant impact on performance, but it is generally better not to compute `Norml2` twice.